commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/auth/DefaultPermissionEvaluator.java

```diff
@@ -60,48 +93,88 @@ public boolean hasPermission(Authentication auth, Object target, Object perm) {
         // check if our entity is shared
         if (target instanceof SharedEntity) {
             SharedEntity sTarget = (SharedEntity) target;
-            switch (perm.toString()) {
-                case "view":
-                case "use":
-                case "copy":
-                case "export":
-                    if (ShareType.PUBLIC.contains(sTarget.getShareType())) {
-                        return true;
-                    }
-                    break;
-                case "save":
-                    if (ShareType.PUBLIC_RW.equals(sTarget.getShareType())) {
+            if (foundPermission != null) {
+                if (foundPermission.isRwOnly() && !ShareType.PUBLIC_RW.equals(sTarget.getShareType())) {
```

Comment:
What is this magical row doing - can you please explain idea behind? This relates to rather not wanted extra column in permissions table.